zilliqa/src/bin/validator_oracle.rs

```diff
@@ -0,0 +1,238 @@
+                loop {
+                    let validators = receiver.borrow_and_update().clone();
+                    if Self::update_validator_manager(&chain_client, &validators)
+                        .await
+                        .is_err()
+                    {
+                        error!(
+                            "Failed updating the validator manager on {}",
+                            chain_client.rpc_url
+                        );
+                    }
+
+                    if receiver.changed().await.is_err() {
+                        break;
+                    }
+                }
+            });
+            handles.push(handle);
+        }
+
+        let result = self.listen_to_staker_updates(sender).await;
+
+        for handle in handles {
+            handle.await?;
+        }
+
+        result
+    }
+
+    async fn listen_to_staker_updates(
+        &mut self,
+        sender: watch::Sender<Vec<NodePublicKey>>,
+    ) -> Result<()> {
+        let filter = Filter::new()
+            .address(contract_addr::DEPOSIT)
+            // Must be the same signature as the event in deposit.sol
+            .event("StakerAdded(bytes)")
+            .from_block(BlockNumberOrTag::Finalized);
+
+        let ws = WsConnect::new(&self.zq2_config.rpc_url);
+        let provider = ProviderBuilder::new().on_ws(ws).await?;
+
+        let subscription = provider.subscribe_logs(&filter).await?;
+        let mut stream = subscription.into_stream();
+        while let Some(log) = stream.next().await {
+            // TODO: infer if staker added or removed
+            info!("Received validator update: {log:?}");
+
+            self.validators = self.get_stakers().await?;
+            info!(
+                "Updating chains to the current validator set: {}\n",
+                serde_json::to_string(&self.validators)?
+            );
+
+            sender.send(self.validators.clone())?;
+        }
+
+        Ok(())
+    }
+
+    async fn create_zq2_chain_client(config: Config, wallet: &LocalWallet) -> Result<ChainClient> {
+        ChainClient::new(
+            &ChainConfig {
+                rpc_url: config.zq2.rpc_url,
+                chain_gateway_address: config.zq2.chain_gateway_address,
+                chain_gateway_block_deployed: 0,
+                block_instant_finality: None,
+                legacy_gas_estimation: None,
+            },
+            H160(**config.zq2.validator_manager_address),
+            wallet.clone(),
+        )
+        .await
+    }
+
+    async fn get_stakers(&self) -> Result<Vec<NodePublicKey>> {
+        debug!("Retreiving validators from the deposit contract");
+
+        let tx = TransactionRequest::new()
+            .to(H160(contract_addr::DEPOSIT.into_array()))
+            .data(contracts::deposit::GET_STAKERS.encode_input(&[])?);
+        // The first chain client is the ZQ2 one.
+        let client = self.chain_clients[0].client.clone();
```

Comment:
This is quite confusing, because in configuration we have a separate entry for the Zilliqa chain but then we merge all the chains together later and have to remember the magic property that Zilliqa is the first one. Why not store it separately from the other chains? i.e. separate the chain you are reading the validator set _from_ and the chains you are sending the validator set _to_.

Reply:
zq2 is really just another chain and shouldn't be treated separately (as far as I can see so far), so I don't think it's too confusing. There should probably be some method that returns it separately, but this is really an implementation detail and shouldn't matter.